### PR TITLE
Add NullableType to types of properties/args that offer it

### PR DESCRIPTION
**Pull request by muglug**
Allows this & similar bugs to be caught: https://github.com/vimeo/psalm/commit/983c6fd6c4e5fbc1b8f18bb496f2a5183ef44614

### CODE REVIEW

File: lib/PhpParser/BuilderAbstract.php [hunk 1]
```diff
@@ -60,7 +60,7 @@ protected function normalizeName($name) {
      * In particular, builtin types are left as strings, custom types become Names and nullables
      * are wrapped in NullableType nodes.
      *
-     * @param Name|string $type The type to normalize
+     * @param Name|string|NullableType $type The type to normalize
```
Comment (nikic) [inline lib/PhpParser/BuilderAbstract.php]: Should be in the `@return` as well

Comment (nikic) [inline lib/PhpParser/BuilderAbstract.php]: A few more builder classes are also affected, e.g. FunctionLike::setReturnType and Param::setTypeHint

Comment (muglug) [inline lib/PhpParser/BuilderAbstract.php]: Thanks! Added those (and a few others)

**Comment by nikic** [comment]
Thanks!

Seeing this type declaration, I'm thinking that PHP-Parser 4 should have some kind of common superclass `Type` with `BuiltinType`, `ClassLikeType` and `NullableType` as children (or similar).

**Comment by muglug** [comment]
Thanks! Right now the class hierarchy is based on structure - all types in docblocks/typehints are handled as `Union` types, even if they only contain a single `Atomic` type. `Atomic` types can be simple, `Generic` or `ObjectLike`. This means support for nullable types is pretty simple - just checking whether a `Union` contains an `Atomic` type with the value `'null'` - a [simple `isset` check](https://github.com/vimeo/psalm/blob/ff409a7f012fb4c8516006c3176d97196060d56b/src/Psalm/Type/Union.php#L135).

Adding support for `NullableType` as a child class of `Union` would mean a non-trivial update to the logic when doing a null check (where a Union type is cloned, then stripped of its null value) and I'm not sure the benefits outweigh.

But if/when do add `NullableType`, I might also add `FalsableType` (bc PHP stdlib loves those possibly`false` return values), and many libraries follow suit.

**Pull request merged**